eqeqeq: an error should point to the equation symbol

The file `a.js`:

``` js
/* global o */
if (o.a == o.b) {
   o.c = o.d;
}
```

gives the following result:

```
a.js
  2:4  error  Expected '===' and instead saw '=='  eqeqeq

✖ 1 problem
```

i.e. it points to the beginning of the whole expression containing the incorrect equality mark. It makes it more difficult to see what's wrong (imagine having something really long instead of `o.a`), it doesn't cooperate with IDE plugins that will mark a perfectly valid `o` token as wrong and it doesn't match jsHint which points to the end of the equation mark.

--- a/lib/rule-context.js
+++ b/lib/rule-context.js
@@ -62,7 +62,7 @@ function RuleContext(ruleId, eslint, severity, options) {
     }, this);
 
     /**
-     * Passthrough to eslint.report() that automatically assigns the rule ID.
+     * Passthrough to eslint.report() that automatically assigns the rule ID and severity.
      * @param {ASTNode} node The AST node related to the message.
      * @param {Object=} location The location of the error.
      * @param {string} message The message to display to the user.

--- a/lib/rules/eqeqeq.js
+++ b/lib/rules/eqeqeq.js
@@ -11,28 +11,88 @@ module.exports = function(context) {
 
     "use strict";
 
+    /**
+     * Checks if either operand of a binary expression is a typeof operation
+     * @param {ASTNode} node The node to check
+     * @returns {boolean} if one of the operands is typeof
+     * @private
+     */
     function isTypeOf(node) {
         return [node.left, node.right].some(function(node) {
             return node.type === "UnaryExpression" && node.operator === "typeof";
         });
     }
 
-    function bothAreSameTypeLiterals(node) {
+    /**
+     * Checks if operands are literals of the same type (via typeof)
+     * @param {ASTNode} node The node to check
+     * @returns {boolean} if operands are of same type
+     * @private
+     */
+    function areLiteralsAndSameType(node) {
         return node.left.type === "Literal" && node.right.type === "Literal" &&
                 typeof node.left.value === typeof node.right.value;
     }
 
+    /**
+     * Checks if operands are both null
+     * @param {ASTNode} node The node to check
+     * @returns {boolean} if operands are null
+     * @private
+     */
     function isNullCheck(node) {
         return (node.right.type === "Literal" && node.right.value === null) ||
                 (node.left.type === "Literal" && node.left.value === null);
     }
 
+    /**
+     * Gets the location (line and column) of the binary expression's operator
+     * @param {ASTNode} node The binary expression node to check
+     * @param {String} operator The operator to find
+     * @returns {Object} { line, column } location of operator
+     * @private
+     */
+    function getLocation(node, operator) {
+        var source, pos, operatorPos, line, column;
+
+        source = context.getSource(node);
+        line = node.loc.start.line;
+        pos = 0;
+
+        // Count lines to end of left expression
+        while (line < node.left.loc.end.line) {
+            pos = source.indexOf("\n", pos) + 1;
+            line += 1;
+        }
+        pos += node.left.loc.end.column;
+
+        operatorPos = source.indexOf(operator, pos);
+
+        // Count lines to operator
+        while ((pos = source.indexOf("\n", pos) + 1) !== 0) {
+            line += 1;
+            if (pos > operatorPos) {
+                line -= 1;
+                break;
+            }
+        }
+
+        // Backtrack to the last newline (if any) to get column
+        column = operatorPos - (source.lastIndexOf("\n", operatorPos) + 1);
+
+        return {
+            line: line,
+            column: column
+        };
+    }
+
     return {
         "BinaryExpression": function(node) {
-            var operator = node.operator;
+            var operator = node.operator,
+                message;
 
             if (context.options[0] === "smart" && (isTypeOf(node) ||
-                    bothAreSameTypeLiterals(node)) || isNullCheck(node)) {
+                    areLiteralsAndSameType(node)) || isNullCheck(node)) {
                 return;
             }
 
@@ -41,10 +101,13 @@ module.exports = function(context) {
             }
 
             if (operator === "==") {
-                context.report(node, "Expected '===' and instead saw '=='.");
+                message = "Expected '===' and instead saw '=='.";
+                context.report(node, getLocation(node, "=="), message);
             } else if (operator === "!=") {
-                context.report(node, "Expected '!==' and instead saw '!='.");
+                message = "Expected '!==' and instead saw '!='.";
+                context.report(node, getLocation(node, "!="), message);
             }
         }
     };
+
 };

--- a/tests/lib/rules/eqeqeq.js
+++ b/tests/lib/rules/eqeqeq.js
@@ -39,6 +39,7 @@ eslintTester.addRuleTest("lib/rules/eqeqeq", {
         { code: "'string' != typeof a", args: [1, "allow-null"], errors: [{ message: "Expected '!==' and instead saw '!='.", type: "BinaryExpression"}] },
         { code: "'hello' != 'world'", args: [1, "allow-null"], errors: [{ message: "Expected '!==' and instead saw '!='.", type: "BinaryExpression"}] },
         { code: "2 == 3", args: [1, "allow-null"], errors: [{ message: "Expected '===' and instead saw '=='.", type: "BinaryExpression"}] },
-        { code: "true == true", args: [1, "allow-null"], errors: [{ message: "Expected '===' and instead saw '=='.", type: "BinaryExpression"}] }
+        { code: "true == true", args: [1, "allow-null"], errors: [{ message: "Expected '===' and instead saw '=='.", type: "BinaryExpression"}] },
+        { code: "a\n==\nb", errors: [{ message: "Expected '===' and instead saw '=='.", type: "BinaryExpression", line: 2 }] }
     ]
 });

